src/mito2/src/sst/parquet/reader.rs

```diff
@@ -249,3 +272,58 @@ impl ParquetReader {
         self.read_format.metadata()
     }
 }
+
+/// Cache layer for parquet's [AsyncFileReader].
+struct AsyncFileReaderCache<T> {
+    /// Underlying async file reader.
+    reader: T,
+    /// Parquet metadata cached locally.
+    metadata: Option<Arc<ParquetMetaData>>,
+    /// Global cache.
+    cache: Option<CacheManagerRef>,
+    /// Id of the region.
+    region_id: RegionId,
+    /// Id of the file to read.
+    file_id: FileId,
+}
+
+impl<T: AsyncFileReader> AsyncFileReader for AsyncFileReaderCache<T> {
+    fn get_bytes(&mut self, range: Range<usize>) -> BoxFuture<'_, Result<Bytes, ParquetError>> {
+        self.reader.get_bytes(range)
+    }
+
+    fn get_byte_ranges(
+        &mut self,
+        ranges: Vec<Range<usize>>,
+    ) -> BoxFuture<'_, Result<Vec<Bytes>, ParquetError>> {
+        self.reader.get_byte_ranges(ranges)
+    }
+
+    fn get_metadata(&mut self) -> BoxFuture<'_, Result<Arc<ParquetMetaData>, ParquetError>> {
+        async {
+            // Tries to get from local cache.
+            if let Some(metadata) = &self.metadata {
+                return Ok(metadata.clone());
+            }
+
+            // Tries to get from global cache.
+            if let Some(metadata) = self
+                .cache
+                .as_ref()
+                .and_then(|cache| cache.get_parquet_meta_data(self.region_id, self.file_id))
+            {
+                return Ok(metadata);
+            }
+
+            // Cache miss.
+            let metadata = self.reader.get_metadata().await?;
+            // Cache the metadata.
+            if let Some(cache) = &self.cache {
+                cache.put_parquet_meta_data(self.region_id, self.file_id, metadata.clone());
+            }
```

Comment:
Why not update the local cache?

Reply:
The builder only calls this method once, not need to update the local cache. The local cache is used to implement a row group level reader in the next PR and will be constructed before creating the async reader. To avoid confusion, I can remove it and add it back later.